share/availability/discovery/set.go

```diff
@@ -47,6 +50,13 @@ func (ps *limitedSet) TryAdd(p peer.ID) error {
 	}
 	if len(ps.ps) < int(ps.limit) {
 		ps.ps[p] = struct{}{}
+
+		// peer will be pushed to the channel only when somebody is reading from it.
+		// this is done to handle case when Peers() was called on empty set.
+		select {
+		case ps.waitPeer <- p:
+		default:
+		}
```

Comment:
Looks like if there are multiple waiters we will only notify one. I suggest for loop with return in default case of select.

Nit: Could refactor it to return error earlier and then.